src/mscorlib/src/System/Threading/Mutex.cs

```diff
@@ -264,20 +151,83 @@ private static OpenExistingResult OpenExistingWorker(string name, MutexRights ri
             }
 
             result = new Mutex(myHandle);
+
             return OpenExistingResult.Success;
         }
 
         // Note: To call ReleaseMutex, you must have an ACL granting you
-        // MUTEX_MODIFY_STATE rights (0x0001).  The other interesting value
+        // MUTEX_MODIFY_STATE rights (0x0001). The other interesting value
         // in a Mutex's ACL is MUTEX_ALL_ACCESS (0x1F0001).
         public void ReleaseMutex()
         {
-            if (Win32Native.ReleaseMutex(safeWaitHandle))
+            if (!Win32Native.ReleaseMutex(safeWaitHandle))
             {
+                throw new ApplicationException(SR.Arg_SynchronizationLockException);
             }
-            else
+        }
+
+        internal class MutexTryCodeHelper
```

Comment:
If you would like, you can get rid of this too and just replace it with try/finally block